### ISSUE TITLE
Weird  issue

### ISSUE THREAD

**Issue by pjebs** [andlabs/ui#220]
I ran go get `github.com/andlabs/ui/` and then ran `go build .` for the code found here: https://github.com/andlabs/ui/wiki/Update-your-UI-in-other-goroutine

The first time I ran it, it worked beautifully.

I clicked the max os x red close button and the window closed. I could see in the terminal that the process shut down correctly.

When I ran the program again (same way as before), the program doesn't run anymore. It displays this error in terminal:

```
 *** Assertion failure in +[NSUndoManager _endTopLevelGroupings], /SourceCache/Foundation/Foundation-1154/Misc.subproj/NSUndoManager.m:340
2017-01-24 18:08:36.541 Test[9371:126254] +[NSUndoManager(NSInternal) _endTopLevelGroupings] is only safe to invoke on the main thread.
2017-01-24 18:08:36.544 Test[9371:126254] (
	0   CoreFoundation                      0x00007fff974af03c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff9a1af76e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff974aee1a +[NSException raise:format:arguments:] + 106
	3   Foundation                          0x00007fff935b999b -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
	4   Foundation                          0x00007fff9353b64f +[NSUndoManager(NSPrivate) _endTopLevelGroupings] + 156
	5   AppKit                              0x00007fff901f9e2e _DPSNextEvent + 2389
	6   AppKit                              0x00007fff901f8e58 -[NSApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 346
	7   AppKit                              0x00007fff901eeaf3 -[NSApplication run] + 594
	8   Test                                0x000000000406cadd uiMain + 29
	9   Test                                0x000000000404e6c0 runtime.asmcgocall + 112
)
2017-01-24 18:08:36.545 Test[9371:126254] *** Assertion failure in +[NSUndoManager _endTopLevelGroupings], /SourceCache/Foundation/Foundation-1154/Misc.subproj/NSUndoManager.m:340
2017-01-24 18:08:36.547 Test[9371:126254] An uncaught exception was raised
2017-01-24 18:08:36.547 Test[9371:126254] +[NSUndoManager(NSInternal) _endTopLevelGroupings] is only safe to invoke on the main thread.
2017-01-24 18:08:36.548 Test[9371:126254] (
	0   CoreFoundation                      0x00007fff974af03c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff9a1af76e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff974aee1a +[NSException raise:format:arguments:] + 106
	3   Foundation                          0x00007fff935b999b -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
	4   Foundation                          0x00007fff9353b64f +[NSUndoManager(NSPrivate) _endTopLevelGroupings] + 156
	5   AppKit                              0x00007fff901eec41 -[NSApplication run] + 928
	6   Test                                0x000000000406cadd uiMain + 29
	7   Test                                0x000000000404e6c0 runtime.asmcgocall + 112
)
2017-01-24 18:08:36.549 Test[9371:126254] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: '+[NSUndoManager(NSInternal) _endTopLevelGroupings] is only safe to invoke on the main thread.'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff974af03c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff9a1af76e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff974aee1a +[NSException raise:format:arguments:] + 106
	3   Foundation                          0x00007fff935b999b -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
	4   Foundation                          0x00007fff9353b64f +[NSUndoManager(NSPrivate) _endTopLevelGroupings] + 156
	5   AppKit                              0x00007fff901eec41 -[NSApplication run] + 928
	6   Test                                0x000000000406cadd uiMain + 29
	7   Test                                0x000000000404e6c0 runtime.asmcgocall + 112
)
libc++abi.dylib: terminating with uncaught exception of type NSException
SIGABRT: abort
PC=0x7fff93bcd286 m=3
signal arrived during cgo execution

goroutine 4 [syscall, locked to thread]:
runtime.cgocall(0x405b120, 0xc420023740, 0x0)
	/usr/local/go/src/runtime/cgocall.go:131 +0x110 fp=0xc420023710 sp=0xc4200236d0
github.com/andlabs/ui._Cfunc_uiMain()
	??:0 +0x41 fp=0xc420023740 sp=0xc420023710
github.com/andlabs/ui.start(0xc420054060, 0x40a1a40)
	/Developer/Projects/GOPATH/src/github.com/andlabs/ui/main.go:56 +0x19b fp=0xc4200237b0 sp=0xc420023740
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc4200237b8 sp=0xc4200237b0
created by github.com/andlabs/ui.Main
	/Developer/Projects/GOPATH/src/github.com/andlabs/ui/main.go:36 +0x6b

goroutine 1 [chan receive]:
github.com/andlabs/ui.Main(0x40a1a40, 0x23, 0x8)
	/Developer/Projects/GOPATH/src/github.com/andlabs/ui/main.go:37 +0xa1
main.main()
	/Users/PJ20/Desktop/Tools/Test/main.go:20 +0x2d

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1

rax    0x0
rbx    0x6
rcx    0xb0102548
rdx    0x0
rdi    0xb03
rsi    0x6
rbp    0xb0102570
rsp    0xb0102548
r8     0x0
r9     0x7fff8fc13d60
r10    0x8000000
r11    0x206
r12    0xb01026d0
r13    0x4461000
r14    0xb0104000
r15    0xb01025b0
rip    0x7fff93bcd286
rflags 0x206
cs     0x7
fs     0x0
gs     0x0
logout

```

**Comment by pjebs** [comment]
Maybe this article is helpful: http://nshipster.com/nsundomanager/

**Comment by pjebs** [comment]
So this is a Go issue? What is the go specific issue so I can file a bug report

**Comment by andlabs** [comment]
I don't know what this is an issue with. I do know that the correct, non-undocumented-call-workaround fix has to be done on the Go side: provide a way to lock a single goroutine and/or its cgo calls to `m0`.

**Comment by pjebs** [comment]
I'm primarily an iOS developer (which is similar to macOS). If we need code to run on the main thread, we can check it using: https://developer.apple.com/reference/foundation/thread and https://developer.apple.com/reference/foundation/thread/1408455-ismainthread

Could that be relevant for testing or atleast checking before running `NSUndoManager`

**Comment by pjebs** [comment]
Also GCD: https://www.raywenderlich.com/60749/grand-central-dispatch-in-depth-part-1 has a function where you can call code guaranteed to be on main thread: `dispatch_get_main_queue()`

**Comment by andlabs** [comment]
You're thinking backwards: it's NSUndoManager that's using `isMainThread` to determine whether it is on the main thread. All of the things you named funnel down to the main CFRunLoop, which can be changed by using an undocumented function exported for the purposes of being usable by kextd; some posts on hte Apple mailing lists have Apple engineers talking about it. I'm already using that function in the package (see `link_darwin_amd64.go`), so I'm not sure why that message is coming back for you.

The problem is that OS X and by extension iOS, watchOS, and tvOS treats "the main thread" and "the first thread created by the OS when you fork() or start the program with Launch Services" as the same thing. On the other hand, on other systems, the main thread can be any thread you want, so long as you don't cross threads. I can spawn a thousand OS threads, then call `CreateWindow()` (Windows) or `gtk_init()` (GTK+) on the 1000th, and everything will still work.

**Comment by pjebs** [comment]
Thanks for the detailed reply.

**Comment by andlabs** [comment]
And are you running ui samples or your own code?

**Comment by pjebs** [comment]
ui samples

**Comment by prsolucoes** [comment]
I think the problem is with UI package:

```
go get github.com/andlabs/ui

ld: warning: object file (/Users/paulo/Developer/workspaces/go/src/github.com/andlabs/ui/libui_darwin_amd64.a(libui-combined.o)) was built for newer OSX version (10.11) than being linked (10.8)
```

**Comment by andlabs** [comment]
That's an unrelated problem with a linking error in libui that will be fixed in the next stable-ish release. If that fix does fix this, then there are things about OS X that I do not understand...

**Comment by prsolucoes** [comment]
Ok, if you can, post here that you update to it test. Ok?

**Comment by childeYin** [comment]
may be you can read this  question https://groups.google.com/forum/#!topic/golang-nuts/E_5-kZnXi5o

**Comment by andlabs** [comment]
Thanks, but that question is rather old; the runtime has changed a lot since then.

I want to propose a `runtime.LockMainOSThread()` or similar, but I'm not sure what the answer to the question "what if two goroutines call this?" would be.

**Comment by andlabs** [comment]
Merged with #280 (which will also explain why the shiny examples worked).

andlabs/libui#308 will fix the built for newer SDK version issue.
